### DISCUSSION TITLE
Gitlab runner failing to execute commands

### DISCUSSION THREAD

**Discussion by gingerbeard72** [batect/batect#1058] [Q&A] [answered]
We have hit a problem with our Gitlab runners where the batect commands are failing with a permission denied error when trying to install make into the container. The following is the condensed output from the runner console for the failed build;

sh batect --no-update-notification --output all validate
validate | Batect | Running validate...
terraform-utils | Batect | Building image...
...
terraform-utils | Image build | #5 [2/2] RUN apk add make
terraform-utils | Image build | #5 0.217 container_linux.go:380: starting container process caused: process_linux.go:545:` container init caused: rootfs_linux.go:58: preparing rootfs caused: permission denied
terraform-utils | Image build | #5 ERROR: executor failed running [/bin/sh -c apk add make]: exit code: 1
terraform-utils | Image build | 

If we run the exact same command when logged into the gitlab runner everything works as expected.

TERRAFORM_WORKSPACE=default SUBFOLDER=my-folder sh batect --no-update-notification --output all validate
validate | Batect | Running validate...
terraform-utils | Batect | Building image...
terraform-utils | Image build | Step 1/2 : FROM hashicorp/terraform:1.0.8
terraform-utils | Image build | ---> a1e27137872c
terraform-utils | Image build | Step 2/2 : RUN apk add make
terraform-utils | Image build | ---> Using cache
terraform-utils | Image build | ---> 8295a8f4d240
terraform-utils | Image build | Successfully built 8295a8f4d240
terraform-utils | Image build | Successfully tagged iac-ec2-pet-servers-terraform-utils:latest

Has anyone run into a similar issue or have any thoughts on why the runner instance has issues executing installation of apps inside the container? 

Docker version info in case this information is of value:
docker version
Client:
Version: 20.10.7
API version: 1.41
Go version: go1.15.14
Git commit: f0df350
Built: Tue Sep 28 19:55:50 2021
OS/Arch: linux/amd64
Context: default
Experimental: true
Server:
Engine:
Version: 20.10.7
API version: 1.41 (minimum version 1.12)
Go version: go1.15.14
Git commit: b0f5bc3
Built: Tue Sep 28 19:56:28 2021
OS/Arch: linux/amd64
Experimental: false
containerd:
Version: 1.4.6
GitCommit: d71fcd7d8303cbf684402823e425e9dd2e99285d
runc:
Version: 1.0.0
GitCommit: 84113eef6fc27af1b01b3181f31bbaf708715301
docker-init:
Version: 0.19.0
GitCommit: de40ad0

**Comment by charleskorn** [comment]
That error (`container_linux.go:380: starting container process caused: process_linux.go:545: container init caused: rootfs_linux.go:58: preparing rootfs caused: permission denied`) is coming from Docker. Could you try running a build of the image directly with Docker as part of your GitLab build to see if we can eliminate Batect as the cause of the issue? Something like `docker build /path/to/image/dir` should be sufficient.

**Comment by gingerbeard72** [comment]
Running docker build - < Dockerfile successfully builds the image.

docker build - < Dockerfile
Sending build context to Docker daemon  2.048kB
Step 1/2 : FROM hashicorp/terraform:1.0.8
1.0.8: Pulling from hashicorp/terraform
a0d0a0d46f8b: Pull complete
07be172ac5f7: Pull complete
37f0c972a6f4: Pull complete
Digest: sha256:c4abe7a8d7b4ae05da852d56093da3704d6684a6d560d31e3a3ab18335cf5dcf
Status: Downloaded newer image for hashicorp/terraform:1.0.8
 ---> a1e27137872c
Step 2/2 : RUN apk add make
 ---> Running in 61013906a42b
fetch https://dl-cdn.alpinelinux.org/alpine/v3.14/main/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.14/community/x86_64/APKINDEX.tar.gz
(1/1) Installing make (4.3-r0)
Executing busybox-1.33.1-r3.trigger
OK: 25 MiB in 32 packages
Removing intermediate container 61013906a42b
 ---> 4b30ee8bc4ce
Successfully built 4b30ee8bc4ce

As mentioned originally, if we run the batect command manually from the server it runs as required. This error is only happening on the gitlab runner servers when we merge the commit. We cannot replicate locally or manually on the instance.

**Reply by charleskorn** [reply]
Just to clarify, is this output from running `docker build` from a build, or directly on the runner?

Also, is your runner configured to run jobs inside a container?

**Reply by gingerbeard72** [reply]
Just to clarify, is this output from running docker build from a build, or directly on the runner?
 - Directly on the runner using a Dockerfile.

Also, is your runner configured to run jobs inside a container?
 - yes they are.

**Reply by charleskorn** [reply]
OK, I suspect there's an issue with running image builds from within the container that GitLab uses for builds. Could you please try adding that `docker build ...` command to your build to see it it works while running inside that container?

**Comment by cmd-werner-diers** [comment, marked as answer]
Hi all, I also came across this issue. In my case this is a docker related issue due to the ***userns-remap*** configuration on the docker daemon.

Versions in use:
- batect: 0.74.0
- docker: 20.10.7 (I've tried with several versions from 18.x, 19.x to 20.x)
- os : amazon linux 2

We have this configuration in the shell runner to ensure all files created by root inside docker belong to gitlab-runner rather than root. 
In other words, this config is mapping root in container to gitlab-runner user on host, so any files created in container for mounted volumes have correct ownership on host.

```shell
  ...
  groupmod -g 100000 gitlab-runner
  usermod -u 100000 -g gitlab-runner

  # subuid: user ID range
  echo "gitlab-runner:100000:1"      >> /etc/subuid
  echo "gitlab-runner:1000000:65536" >> /etc/subuid
  # subgid: group ID range
  echo "gitlab-runner:100000:1"      >> /etc/subgid
  echo "gitlab-runner:1000000:65536" >> /etc/subgid
  
  ...
  cat  /etc/docker/daemon.json
  {
      "userns-remap": "gitlab-runner"
  }
```

Note that this configuration works well with `3 Musketeers Pattern` but causes issues with `batect` (we are testing both approaches in the same gitlab runner using shell executor).
As it was already mentioned, if I log in as gitlab-runner on the gitlab runner (EC2) and then run manually the commands on the same folder as the failed job was executed, everything works.


From the system logs, after gitlab-runner fetches the job queue:
```shell
  8548 Oct 12 04:06:47 ip-10-129-56-163 gitlab-runner: Checking for jobs... received #033[0;m  job#033[0;m=1670275843 repo_url#033[0;m=https://gitlab.com/my-project.git runner#033[0;m=nh8fc1_2
  8560 Oct 12 04:06:49 ip-10-129-56-163 dockerd: time="2021-10-12T04:06:49.556233400Z" level=warning msg="failed to retrieve runc version: unknown output format: runc version 1.0.0\ncommit: 84113eef6fc27af1b01b3181f31bbaf708715301\nsp     ec: 1.0.2-dev\ngo: go1.15.14\nlibseccomp: 2.4.1\n"
  8584 Oct 12 04:07:22 ip-10-129-56-163 dockerd: Unsolicited response received on idle HTTP channel starting with "HTTP/1.0 408 Request Time-out\r\ncache-control: no-cache\r\ncontent-type: text/html\r\n\r\n<html><body><h1>408 Request      Time-out</h1>\nYour browser didn't send a complete request in time.\n</body></html>\n\n"; err=<nil>
```

When building docker images with batect we encountered the following error:
```shell
  error 0.207 container_linux.go:380: starting container process caused: process_linux.go:545: container init caused: rootfs_linux.go:58: preparing rootfs caused: permission denied
```


Another error I see is:
```shell
  Starting container 'df9f287' failed: OCI runtime create failed: container_linux.go:380: starting container process caused: setup user: cannot set uid to unmapped user in user namespace: unknown
```

Which I have potentially fixed by expanding the range on both files `subuid` and `subgid`. 
From `gitlab-runner:1000000:65536` to `gitlab-runner:1000000:165536`

For example :
```shell
  lint:helm | Batect | Running lint:helm...
  helm      | Batect | Building image...
  helm      | Image build | #1 [internal] load build definition from Dockerfile
  helm      | Image build | #1 ...
  helm      | Image build | 
  helm      | Image build | #2 [internal] load .dockerignore
  helm      | Image build | #2 ...
  helm      | Image build | 
  ...
  helm      | Image build | #4 [1/2] FROM docker.io/alpine/helm:3.7.0@sha256:f583006048a07f2bcb5467c893374cc93dff866149e653226e14ede325b0a5d1
  helm      | Image build | #4 DONE
  helm      | Image build | 
  helm      | Image build | #5 [2/2] RUN apk add jq &&     helm plugin install --version v0.10.0 https://github.com/chartmuseum/helm-push.git
  helm      | Image build | #5 CACHED
  helm      | Image build | 
  helm      | Image build | #6 exporting to image
  helm      | Image build | #6 exporting layers: done
  helm      | Image build | #6 writing image sha256:c6f07fd9f4835aa8fdb21c664341c6e1379667f42bb6a26dab68baf4c1b4e142: done
  helm      | Image build | #6 naming to docker.io/library/ci-helm: done
  helm      | Image build | #6 DONE
  helm      | Image build | 
  helm      | Batect | Image built.
  helm      | Batect | Running " helm lint helm/${CHART_NAME} --values test-values.yml --with-subcharts ${*} " --...
  helm      ! Batect | Error: Could not run container helm.
            ! Batect | Starting container 'df9f287' failed: OCI runtime create failed: container_linux.go:380: starting container process caused: setup user: cannot set uid to unmapped user in user namespace: unknown
  lint:helm | Batect | Cleaning up...
  lint:helm ! Batect | The task lint:helm failed. See above for details.
```

**Reply by charleskorn** [reply]
Do you have `run_as_current_user` enabled in your project @cmd-werner-diers? You should only need one or the other (user namespace remapping), and I suspect having both enabled would produce the error you're seeing (`starting container process caused: setup user: cannot set uid to unmapped user in user namespace: unknown`)
